Add support for Asp.Net controller for Query

ViniciusSouza opened the pull request:
The main go of this PR is adding the Asp.Net controller to Arriba Application Query Routes, during the process some refactors was performed:

- REFACTOR: Separate Monitor contract to ITelemetry
- TEST: Refactoring ArribaManagementServiceTests 
- REFACTOR: Add ITelemetry to IArribaManagementService

ViniciusSouza committed:
FEATURE: Check for read permission

ViniciusSouza committed:
REFACTOR: Separate Service from Response calls

ViniciusSouza committed:
FEATURE: Translate Exceptions into Responses

ViniciusSouza committed:
REFACTOR: Extract IArribaQueryService

ViniciusSouza committed:
TEST: Refactoring ArribaManagementServiceTests

Creating a ArribaServiceBase to be used as a base class for Management and Query services

ViniciusSouza committed:
REFACTOR: Decouple ArribaQueryServices from ArribaQueryApplication

ViniciusSouza committed:
FEATURE: Use ParamChecker guard methods

ViniciusSouza committed:
TEST: Unit test for ArribaQueryServices

ViniciusSouza committed:
FIX: Initializing ArribaServices

ViniciusSouza committed:
FIX: Using Newtonsoft.Json as the default serializer

By default the AspNet.Core Serializer on 3.0+ is the System.Text.Json, the current implementation of the library doesn't support Fields only Properties even public fields.

There is a change that will be part of .net 5.0 that will provide a option to change this behaviour
https://github.com/dotnet/runtime/pull/36986

ericmaino commented on src/Arriba/Arriba/Monitoring/ITelemetry.cs:
```diff
@@ -0,0 +1,22 @@
+        /// <returns>A disposable handle.</returns>
+        IDisposable Monitor(MonitorEventLevel level, string name, string type = null, string identity = null, object detail = null);
+
+        List<MonitorEventScope> MonitorEvents();
```
Try to remember to use contracts on interfaces, not concrete types. I would have also made this a Get Only property. Example below

```
IList<MonitorEventScope> MonitorEvents {get;}
```

ViniciusSouza replied:
I had missed that thanks!

ericmaino commented on src/Arriba/Arriba/Server/Application/ArribaQueryApplication.cs:
```diff
@@ -448,24 +476,24 @@ private async Task<IResponse> Aggregate(IRequestContext ctx, Route route)
-            string aggregationFunction = p["a"] ?? "count";
-            string columnName = p["col"];
-            string queryString = p["q"];
+            string aggregationFunction = parameters["a"] ?? "count";
```
Little things like leaving parameters as p for this refactoring would help to reduce some of the noise

ViniciusSouza replied:
Did you mean in all the changes where I renamed p to parameters? 
And I though that having a more meaningful name would help read the code

ericmaino replied:
The rename is good, but it should not be combined with a code move like this. They should be separate refactorings as then it's very clear what each is.

ericmaino commented on src/Arriba/Arriba/Server/Application/ArribaApplication.cs:
```diff
@@ -240,5 +240,18 @@ protected override IResponse OnAfterProcess(IRequestContext request, IResponse r
             return response;
         }
+
+        protected IResponse ExceptionToArribaResponse(Exception ex)
```
It's hard to understand why this was completed as the next commit doesn't seem to take advantage of this. Might have been useful to combine it with the commit that required this move or to ensure it's the next commit in the stack

ericmaino commented on src/Arriba/Arriba/Server/Application/ArribaQueryApplication.cs:
```diff
@@ -345,72 +298,6 @@ private static ByteBlock ConvertToByteBlock(object value)
             }
         }
 
-        private AllCountResult AllCount(ITelemetry telemetry, NameValueCollection parameters, IPrincipal user)
```
Reordering the code plus making changes makes this a bit harder to follow and review

ericmaino commented on src/Arriba/Arriba/Server/Application/ArribaQueryApplication.cs:
```diff
@@ -35,16 +35,16 @@ public ArribaQueryApplication(DatabaseFactory f, ClaimsAuthenticationService aut
             _service = queryServices;
 
             // /table/foo?type=select
-            this.GetAsync(new RouteSpecification("/table/:tableName", new UrlParameter("action", "select")), this.ValidateReadAccessAsync, this.Select);
```
My assumption is that this has been removed because the service now handles this instead of having to decorate each one. An additional comment in the request saying where to look to see this is still being performed would be valuable.

or something like REFACTOR: Remove duplicate route chain validation

ViniciusSouza replied:
Do you think that combining the commits, will help on getting a better understanding?

FEATURE: Check for read permission
REFACTOR: Remove duplicate route chain validation

ericmaino replied:
Yes, if the commits are named similarly that would be preferred

FEATURE: Checking read permissions in the service
FIX: Removing duplicate read permission check

ericmaino commented on src/Arriba/Arriba.Server/Filters/ArribaResponseFilter.cs:
```diff
@@ -13,7 +13,7 @@ public ArribaResultFilter()
         public override void OnResultExecuting(ResultExecutingContext context)
         {
-            if (context.Controller is ArribaController)
+            if (context.Controller is ArribaManagementController)
```
Does the envelop not need to apply to both the Management controller and the Query controller?

ViniciusSouza replied:
They are all ArribaController as I'm using partial class

ericmaino commented on src/Arriba/Arriba.Server/Controllers/ArribaManagementController.cs:
```diff
@@ -16,10 +17,12 @@ namespace Arriba.Controllers
     public class ArribaManagementController : ControllerBase
     {
         private readonly IArribaManagementService _arribaManagement;
+        private readonly ITelemetry _telemetry;
```
Feels like we should pull this commit out of the changes and take it separately. It feels pretty invasive and not sure if it's the way we'd like to align on this change, yet

ericmaino commented on src/Arriba/Arriba.Server/Controllers/ArribaActionController.cs:
```diff
@@ -0,0 +1,39 @@
+
+namespace Arriba.Controllers
+{
+    public partial class ArribaController
```
If we wish to go the partial class route I'd recommend we follow a naming conventions such as ArribaContoller.cs and ArribaController.Query.cs

ViniciusSouza committed:
REFACTOR: Separate Monitor contract to ITelemetry

ViniciusSouza committed:
FEATURE: Add ArribaQueryController

ViniciusSouza committed:
FEATURE: Support for ArribaRoutes V1

Some Rest API routes was not well formed, in order to move forward to V2 without breaking the UI
was added a ArribaActionController that demanded a Refactor on Arriba's controllers
